File: fsharp/relative-distance/RelativeDistance.fs
module RelativeDistance

type FamilyTree = Map<string, string list>
type Graph = Map<string, Set<string>>

let buildGraph (familyTree : FamilyTree) : Graph =
    let addEdge source target =
        Map.change source (Option.defaultValue Set.empty >> Set.add target >> Some)
    
    familyTree
    |> Map.fold (fun graph parent children -> // use a mutable graph variable to many edges at once
        let mutable g = graph
        for child in children do
            g <- addEdge parent child g
            g <- addEdge child parent g
        for c1 in children do // sibling connections
            for c2 in children do
                if c1 <> c2 then 
                    g <- addEdge c1 c2 g
        g
    ) Map.empty

let degreeOfSeparation (familyTree : FamilyTree) (person1 : string) (person2 : string) : int option =
    let graph = buildGraph familyTree
    let rec bfs queue visited =
        match queue with
        | [] -> None // no connection found, queue empty
        | (current, dist) :: _ when current = person2 -> Some dist
        | (current, dist) :: rest ->
            let nextNodes =
                Map.tryFind current graph
                |> Option.defaultValue Set.empty
                |> Set.filter (fun n -> not (Set.contains n visited))
                |> Set.map (fun n -> (n, dist + 1))
                |> Set.toList
            bfs (rest @ nextNodes) (Set.add current visited)
    
    bfs [(person1, 0)] Set.empty

File: csharp/relative-distance/RelativeDistance.cs
public class RelativeDistance
{
    private readonly Dictionary<string, HashSet<string>> _graph;

    public RelativeDistance(Dictionary<string, string[]> familyTree)
    {
        _graph = new Dictionary<string, HashSet<string>>();
        foreach (var (parent, children) in familyTree)
        {
            foreach (var child in children)
            {
                AddBidirectionalEdge(parent, child);
                foreach (var sibling in children)
                {
                    if (sibling != child)
                    {
                        AddBidirectionalEdge(child, sibling);
                    }
                }
            }
        }
    }

    private void AddBidirectionalEdge(string p1, string p2)           
    {
        if (!_graph.ContainsKey(p1))
        {
            _graph[p1] = new HashSet<string>();
        }
        if (!_graph.ContainsKey(p2))
        {
            _graph[p2] = new HashSet<string>();
        }
        _graph[p1].Add(p2);
        _graph[p2].Add(p1);
    }

    public int DegreeOfSeparation(string personA, string personB)
    {
        if (!_graph.ContainsKey(personA) || !_graph.ContainsKey(personB))
        {
            return -1;
        }

        var visited = new HashSet<string>();
        var queue = new Queue<(string person, int dist)>();
        queue.Enqueue((personA, 0));
        visited.Add(personA);

        while (queue.Count > 0)
        {
            var (currentPerson, dist) = queue.Dequeue();
            if (currentPerson == personB)
            {
                return dist;
            }
            if (_graph.TryGetValue(currentPerson, out var neighbors))
            {
                foreach (var neighbor in neighbors)
                {
                    if (visited.Add(neighbor))
                    {
                        queue.Enqueue((neighbor, dist + 1));
                    }
                }
            }
        }
        return -1; // no connection found
    }
}
